File: app/src/main/java/com/example/mutidemo/ui/GCJ02ToWGS84Activity.java
package com.example.mutidemo.ui;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.location.Location;
import android.util.Log;

import com.amap.api.location.AMapLocation;
import com.esri.arcgisruntime.geometry.Point;
import com.esri.arcgisruntime.geometry.SpatialReference;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.BasemapStyle;
import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
import com.esri.arcgisruntime.symbology.PictureMarkerSymbol;
import com.esri.arcgisruntime.util.ListenableList;
import com.example.mutidemo.R;
import com.example.mutidemo.base.AndroidxBaseActivity;
import com.example.mutidemo.databinding.ActivityGisBinding;
import com.example.mutidemo.util.LocationHelper;
import com.example.mutidemo.util.OtherUtils;
import com.example.mutidemo.util.callback.IAddressListener;
import com.example.mutidemo.util.callback.ILocationListener;

public class GCJ02ToWGS84Activity extends AndroidxBaseActivity<ActivityGisBinding> {

    private static final String TAG = "GCJ02ToWGS84Activity";
    private final Context context = this;

    @Override
    public void initData() {
        viewBinding.mapView.setAttributionTextVisible(false);//去掉左下角属性标识
        viewBinding.mapView.setViewpointScaleAsync(2800);//数字越大，放大比例越小，缩放比例[36000,250]

        ArcGISMap arcGISMap = new ArcGISMap(BasemapStyle.ARCGIS_STREETS);
        viewBinding.mapView.setMap(arcGISMap);
    }

    @Override
    public void initEvent() {
        LocationHelper.obtainCurrentLocationByGD(this, new ILocationListener() {
            @Override
            public void onLocationGet(Location location) {

            }

            @Override
            public void onAMapLocationGet(AMapLocation aMapLocation) {
                if (aMapLocation != null) {
                    double[] gcjToWgs = LocationHelper.gcjToWgs(aMapLocation.getLongitude(), aMapLocation.getLatitude());
                    Log.d(TAG, "GCJ-02: [" + gcjToWgs[0] + "," + gcjToWgs[1] + "]");
                    Point point = new Point(gcjToWgs[0], gcjToWgs[1], SpatialReference.create(4326));
                    addPictureMarker(point, false);
                    viewBinding.mapView.setViewpointCenterAsync((point), 28000);
                }
            }
        }, true);

        viewBinding.expandMapView.setOnClickListener(view -> viewBinding.mapView.setViewpointScaleAsync(viewBinding.mapView.getMapScale() * 0.5));
        viewBinding.minusMapView.setOnClickListener(view -> viewBinding.mapView.setViewpointScaleAsync(viewBinding.mapView.getMapScale() * 2));
        viewBinding.removeToLocalView.setOnClickListener(view -> {
            OtherUtils.showLoadingDialog(GCJ02ToWGS84Activity.this, "定位中，请稍后");
            LocationHelper.obtainCurrentLocation(GCJ02ToWGS84Activity.this, new ILocationListener() {
                @Override
                public void onLocationGet(Location location) {
                    if (location != null) {
                        Log.d(TAG, "WGS-84: [" + location.getLongitude() + "," + location.getLatitude() + "]");
                        Point point = new Point(location.getLongitude(), location.getLatitude(), SpatialReference.create(4326));
                        addPictureMarker(point, true);
                        viewBinding.mapView.setViewpointCenterAsync((point), 28000);
                        OtherUtils.dismissLoadingDialog();
                        //显示具体位置
                        LocationHelper.antiCodingLocation(context, location.getLongitude(), location.getLatitude(), new IAddressListener() {
                            @Override
                            public void onGetAddress(String address) {
                                viewBinding.addressView.setText(address);
                            }
                        });
                    }
                }

                @Override
                public void onAMapLocationGet(AMapLocation aMapLocation) {

                }
            }, true);
        });
    }

    private void addPictureMarker(Point point, boolean isGPS) {
        Bitmap caseBitmap;
        if (isGPS) {
            caseBitmap = BitmapFactory.decodeResource(getResources(), R.mipmap.location_handle);
        } else {
            caseBitmap = BitmapFactory.decodeResource(getResources(), R.mipmap.location);
        }
        BitmapDrawable caseDrawable = new BitmapDrawable(getResources(), caseBitmap);
        PictureMarkerSymbol pictureMarker = new PictureMarkerSymbol(caseDrawable);
        pictureMarker.setWidth(24);
        pictureMarker.setHeight(24);
        pictureMarker.loadAsync();

        Graphic graphic = new Graphic(point, pictureMarker);
        GraphicsOverlay mGraphicsOverlay = new GraphicsOverlay();
        ListenableList<Graphic> overlayGraphics = mGraphicsOverlay.getGraphics();
        ListenableList<GraphicsOverlay> graphicsOverlays = viewBinding.mapView.getGraphicsOverlays();
        overlayGraphics.add(graphic);
        graphicsOverlays.add(mGraphicsOverlay);
    }

    protected void onResume() {
        super.onResume();
        viewBinding.mapView.resume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        viewBinding.mapView.pause();
    }

    @Override
    public void onDestroy() {
        viewBinding.mapView.dispose();
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/mutidemo/view/GCJ02ToWGS84Activity.kt
package com.example.mutidemo.view

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.location.Location
import android.os.Bundle
import android.util.Log
import com.amap.api.location.AMapLocation
import com.amap.api.services.core.AMapException
import com.esri.arcgisruntime.geometry.Point
import com.esri.arcgisruntime.geometry.SpatialReference
import com.esri.arcgisruntime.mapping.ArcGISMap
import com.esri.arcgisruntime.mapping.BasemapStyle
import com.esri.arcgisruntime.mapping.view.Graphic
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay
import com.esri.arcgisruntime.symbology.PictureMarkerSymbol
import com.example.mutidemo.R
import com.example.mutidemo.callback.IAddressListener
import com.example.mutidemo.callback.ILocationListener
import com.example.mutidemo.util.LoadingDialogHub
import com.example.mutidemo.util.LocationHelper
import com.pengxh.kt.lite.base.KotlinBaseActivity
import kotlinx.android.synthetic.main.activity_gis.*

class GCJ02ToWGS84Activity : KotlinBaseActivity() {

    private val kTag = "GCJ02ToWGS84Activity"
    private val context: Context = this@GCJ02ToWGS84Activity

    override fun setupTopBarLayout() {}

    override fun observeRequestState() {

    }

    override fun initLayoutView(): Int = R.layout.activity_gis

    override fun initData(savedInstanceState: Bundle?) {
        mapView.isAttributionTextVisible = false //去掉左下角属性标识
        mapView.setViewpointScaleAsync(2800.0) //数字越大，放大比例越小，缩放比例[36000,250]
        val arcGISMap = ArcGISMap(BasemapStyle.ARCGIS_STREETS)
        mapView.map = arcGISMap
    }

    override fun initEvent() {
        LocationHelper.obtainCurrentLocationByGD(this, object : ILocationListener {
            override fun onLocationGet(location: Location?) {}
            override fun onAMapLocationGet(aMapLocation: AMapLocation?) {
                if (aMapLocation != null) {
                    val gcjToWgs = LocationHelper.gcjToWgs(
                        aMapLocation.longitude, aMapLocation.latitude
                    )
                    Log.d(kTag, "GCJ-02: [" + gcjToWgs[0] + "," + gcjToWgs[1] + "]")
                    val point: Point = Point(
                        gcjToWgs[0], gcjToWgs[1], SpatialReference.create(4326)
                    )
                    addPictureMarker(point, false)
                    mapView.setViewpointCenterAsync(point, 28000.0)
                }
            }
        }, true)
        expandMapView.setOnClickListener {
            mapView.setViewpointScaleAsync(
                mapView.mapScale * 0.5
            )
        }
        minusMapView.setOnClickListener {
            mapView.setViewpointScaleAsync(
                mapView.mapScale * 2
            )
        }
        removeToLocalView.setOnClickListener {
            LoadingDialogHub.show(this, "定位中，请稍后")
            LocationHelper.obtainCurrentLocation(
                this, object : ILocationListener {
                    override fun onLocationGet(location: Location?) {
                        if (location != null) {
                            Log.d(
                                kTag,
                                "WGS-84: [" + location.longitude + "," + location.latitude + "]"
                            )
                            val point = Point(
                                location.longitude, location.latitude, SpatialReference.create(4326)
                            )
                            addPictureMarker(point, true)
                            mapView.setViewpointCenterAsync(point, 28000.0)
                            LoadingDialogHub.dismiss()
                            //显示具体位置
                            try {
                                LocationHelper.antiCodingLocation(
                                    context, location.longitude, location.latitude,
                                    object : IAddressListener {
                                        override fun onGetAddress(address: String?) {
                                            addressView.text = address
                                        }
                                    })
                            } catch (e: AMapException) {
                                e.printStackTrace()
                            }
                        }
                    }

                    override fun onAMapLocationGet(aMapLocation: AMapLocation?) {}
                },
                true
            )
        }
    }

    private fun addPictureMarker(point: Point, isGPS: Boolean) {
        val caseBitmap = if (isGPS) {
            BitmapFactory.decodeResource(resources, R.mipmap.location_handle)
        } else {
            BitmapFactory.decodeResource(resources, R.mipmap.location)
        }
        val caseDrawable = BitmapDrawable(resources, caseBitmap)
        val pictureMarker = PictureMarkerSymbol(caseDrawable)
        pictureMarker.width = 24f
        pictureMarker.height = 24f
        pictureMarker.loadAsync()
        val graphic = Graphic(point, pictureMarker)
        val mGraphicsOverlay = GraphicsOverlay()
        val overlayGraphics = mGraphicsOverlay.graphics
        val graphicsOverlays = mapView.graphicsOverlays
        overlayGraphics.add(graphic)
        graphicsOverlays.add(mGraphicsOverlay)
    }

    override fun onResume() {
        super.onResume()
        mapView.resume()
    }

    override fun onPause() {
        super.onPause()
        mapView.pause()
    }

    override fun onDestroy() {
        mapView.dispose()
        super.onDestroy()
    }
}